### PR TITLE
fix(runtime): replaceSync usage matches webAPI spec

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Unit tests (`npm test`) were run locally and passed
- [ ] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

When SSR/mock environments like Happy DOM monkey-patch the [CSSStyleSheet API](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/replaceSync) and invoke `replaceSync` an exception can be thrown. Happy DOM expects the CSS text to be defined, but when `replaceSync` is called from Stencil, the contents can be `undefined`. 

The behavior regressed in this commit: https://github.com/ionic-team/stencil/commit/c47cec6581d4409e8261b3516b78532b5c49d079

GitHub Issue Number: N/A

## What is the new behavior?

- Updates the usage of `replaceSync` to pass an empty string when the CSS text contents are undefined.
- Stencil web components can be tested with Happy DOM (and likely other mock environments with this monkey-patch pattern).

The proposed changes have zero impact on existing Stencil developers. The WebAPI implementation of `replaceSync` returns the same result when invoked with: `replaceSync(undefined)` or `replaceSync('')` (both return `undefined`). 

The WebAPI spec has this note regarding the `text` parameter of `replaceSync`:
> A string containing the style rules to replace the content of the stylesheet. If the string does not contain a parseable list of rules, then the value will be set to an empty string.

This makes me think that this change is appropriate to Stencil instead of making a change to Happy DOM's internals. 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

- Created an Ionic React Starter
- Installed [`@happy-dom/jest-environment`](https://www.npmjs.com/package/@happy-dom/jest-environment)
- Updated the `test` script to assign the jest environment
- Install Ionic Dev build (due to this issue: https://github.com/ionic-team/ionic-framework/issues/25900)
```
npm install @ionic/react@6.2.7-dev.11662662897.1987afe4 @ionic/react-router@6.2.7-dev.11662662897.1987afe4
```
- Run the test suite
- Observed: Exception is thrown for trying to perform `.replace` on `undefined`:
```
/Users/sean/Documents/ionic/issues/react-happy-dom/node_modules/happy-dom/lib/css/CSSParser.js:25
        const css = cssText.replace(COMMENT_REGEXP, '');
                            ^

TypeError: Cannot read properties of undefined (reading 'replace')
    at Function.parseFromString (/Users/sean/Documents/ionic/issues/react-happy-dom/node_modules/happy-dom/src/css/CSSParser.ts:23:23)
    at CSSStyleSheet.replaceSync (/Users/sean/Documents/ionic/issues/react-happy-dom/node_modules/happy-dom/src/css/CSSStyleSheet.ts:129:3)
    at registerStyle (/Users/sean/Documents/ionic/issues/react-happy-dom/node_modules/@stencil/core/internal/client/index.js:233:19)
```

- Manually updated the compiled output to the changes proposed in this PR
- Re-ran test suite
- Observed: No exception is raised

### CODE REVIEW

**Comment by rwaskiewicz** [comment]
I want to dig into this a little bit more tomorrow to understand the path where we're allowing `undefined` for `cssText`, because our function signature is currently:
```
export const registerStyle = (scopeId: string, cssText: string, allowCS: boolean) => {
```
which may not be correct.  I did run a quick `strictNullChecks` test and found
```
src/runtime/bootstrap-lazy.ts:48:21 - error TS2345: Argument of type 'string | null' is not assignable to parameter of type 'string'.
  Type 'null' is not assignable to type 'string'.

48       registerStyle(styles[i].getAttribute(HYDRATED_STYLE_ID), convertScopedToShadow(styles[i].innerHTML), true);
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
which I can pull on a little bit more tomorrow.

I also wanna TAL at the spec, because IIRC, both `replace` and `replaceSync` should be able to turn an unparsable set of CSS rules into an empty list.

tl;dr - thanks and stay tuned!

**Comment by rwaskiewicz** [comment]
@sean-perkins I wanna make sure I'm reproducing this correctly and wanted to confirm a few things:
1. Project initialization - I init'ed my reproduction project using `ionic start`, selected 'React' as my framework, and opened a blank project - does that sound about right?
2. Environment setup
After running `npm i -D @happy-dom/jest-environment`, I configured the `test` script in `package.json` like so:
```diff
diff --git a/package.json b/package.json
index 0276169..9403d10 100644
--- a/package.json
+++ b/package.json
@@ -43,7 +43,7 @@
   "scripts": {
     "start": "react-scripts start",
     "build": "react-scripts build",
-    "test": "react-scripts test --transformIgnorePatterns 'node_modules/(?!(@ionic/react|@ionic/react-router|@ionic/core|@stencil/core|ionicons)/)'",
+    "test": "react-scripts test --transformIgnorePatterns 'node_modules/(?!(@ionic/react|@ionic/react-router|@ionic/core|@stencil/core|ionicons)/)' --env=@happy-dom/jest-environment",
     "eject": "react-scripts eject"
   },
   "eslintConfig": {
```
Does that look right?

3. Finally, test running - I ran `npm t`, but got a different error:
```
 RUNS  src/App.test.tsx
/Users/ryan/workspaces/sandbox/stencil-issues/happydom/my-hd-test/node_modules/@ionic/core/components/dir.js:24
  return (document === null || document === void 0 ? void 0 : document.dir.toLowerCase()) === 'rtl';
                                                                           ^

TypeError: Cannot read properties of undefined (reading 'toLowerCase')
    at isRTL (/Users/ryan/workspaces/sandbox/stencil-issues/happydom/my-hd-test/node_modules/@ionic/core/components/dir.js:24:76)
    at Object.createSwipeBackGesture (/Users/ryan/workspaces/sandbox/stencil-issues/happydom/my-hd-test/node_modules/@ionic/core/components/swipe-back.js:19:26)
    at Object.<anonymous>.RouterOutlet.mode.connectedCallback (/Users/ryan/workspaces/sandbox/stencil-issues/happydom/my-hd-test/node_modules/@ionic/core/components/ion-router-outlet.js:67:124)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```
which _may_ be related to my first question regarding the starter used, potentially question two as well.

Either way, LMK and we'll keep going on this issue

**Comment by sean-perkins** [comment]
@rwaskiewicz your reproduction steps are correct.

The error you are receiving is from an Ionic Framework bug [fixed here](https://github.com/ionic-team/ionic-framework/pull/25901).

If the additional exceptions do not log, you can try installing the dev-build:

```bash
npm install @ionic/react@6.2.7-dev.11662662897.1987afe4 @ionic/react-router@6.2.7-dev.11662662897.1987afe4
```

This should alleviate the first exception and allow the Stencil related one to fail next. I'll update the reproduction steps 👍

**Comment by sean-perkins** [comment]
@rwaskiewicz strange.. only happens in a new spec file. 

For example, update your reproduction to have `Home.test.tsx` in the `src/pages/` directory:

```ts
import React from "react";
import { render } from "@testing-library/react";
import Home from "./Home";

test("renders without crashing", () => {
  const { baseElement } = render(<Home />);
  expect(baseElement).toBeDefined();
});
```

```
/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/happy-dom/lib/css/CSSParser.js:25
        const css = cssText.replace(COMMENT_REGEXP, '');
                            ^

TypeError: Cannot read properties of undefined (reading 'replace')
    at Function.parseFromString (/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/happy-dom/src/css/CSSParser.ts:23:23)
    at CSSStyleSheet.replaceSync (/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/happy-dom/src/css/CSSStyleSheet.ts:128:42)
    at registerStyle (/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/@stencil/core/internal/client/index.js:330:13)
    at initializeComponent (/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/@stencil/core/internal/client/index.js:2843:9)
    at connectedCallback (/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/@stencil/core/internal/client/index.js:2966:9)
    at Object.<anonymous>.Toolbar.color.connectedCallback (/Users/sean/Documents/ionic/issues/stencil-replacesync-reproduction/node_modules/@stencil/core/internal/client/index.js:3070:7)
```

**Comment by rwaskiewicz** [comment]
Thanks! I was able to reproduce this with a new test file. 

FWIW, I think it has something to do with the contents of `Home.tsx` - if I put the contents of `src/App.tsx` in `src/pages/Home.tsx` and fixup the 'themes.css' import statement, I don't get the issue, which is odd...

But anyway, in the summary of the PR we have:

> The WebAPI spec has this note regarding the text parameter of replaceSync:

>>    A string containing the style rules to replace the content of the stylesheet. If the string does not contain a parseable list of rules, then the value will be set to an empty string.

> This makes me think that this change is appropriate to Stencil instead of making a change to Happy DOM's internals.


Can you expound on that last line a little for me please? I think I would argue that the mock API should be responsible for following the spec, rather than the calling library - and I say that knowing the pain that is maintaining Stencil's mock doc environment 😆 But maybe I'm not considering something?

**Comment by sean-perkins** [comment]
This bit in particular:

> If the string does not contain a parseable list of rules, then the value will be set to an empty string.

In Stencil, the value being passed to `replaceSync` is `undefined`, not a string nor an empty string when it does not contain a parseable list of rules. 

This is problematic for implementations that monkey-patch the API, as it is receiving invalid/unexpected inputs. Happy DOM could check for undefined and do nothing, but since they aren't having to do that for Lit, React, Angular or Vue, it feels like a Stencil problem.

**Comment by rwaskiewicz** [comment]
Ah I see. I think the breakdown here is that I interpret `undefined` _as_ an unparseable list of rules. Testing this in Chrome/Edge/Firefox, they each handle `undefined` and `null` without crashing:
```html
<html>

<body>
    <div id="my-div"></div>

    <script>
        const div = document.querySelector('#my-div');
        const shadow = div.attachShadow({ mode: 'open' });

        // Create a new CSSStyleSheet, we'll reference this later
        const sheet = new CSSStyleSheet();
        shadow.adoptedStyleSheets = [sheet];

        // This is the span that will be styled
        const span = document.createElement('span');
        span.textContent = "Hello World";

        // add the span to the shadow dom for styling to be applied
        shadow.appendChild(span);

        // after 2 seconds, change the style
        const timer = setTimeout(() => {
            const style = undefined; // 'span {color: red}';
            sheet.replaceSync(style);
            // added this so that we can verify this fired
            console.log(`Replaced style with ${style}`);
            clearTimeout(timer);
        }, 2000);
    </script>
</body>

</html>
```

(Safari has issues creating a `CSSStyleSheet` without any args 😆 😭)

**Comment by sean-perkins** [comment]
Discussed with the team - we are going to close out this PR. If you require this change or have a similar problem, please open a new issue and the team will take a fresh look at it. Thanks!
